File: src/core/browser.py
"""
브라우저 관리 (Playwright Browser Manager)
"""
import asyncio
from typing import Optional
from playwright.async_api import async_playwright, Browser, BrowserContext, Page, Playwright
from tenacity import retry, stop_after_attempt, wait_exponential

from .config import SiteConfig
from ..utils.logger import get_logger

logger = get_logger()


class BrowserManager:
    """Playwright 브라우저 관리"""
    
    def __init__(self, config: SiteConfig):
        self.config = config
        self._playwright: Optional[Playwright] = None
        self._browser: Optional[Browser] = None
        self._context: Optional[BrowserContext] = None
        self._page: Optional[Page] = None
    
    async def __aenter__(self):
        """비동기 컨텍스트 매니저 진입"""
        await self.start()
        return self
    
    async def __aexit__(self, exc_type, exc_val, exc_tb):
        """비동기 컨텍스트 매니저 종료"""
        await self.close()
    
    async def start(self):
        """브라우저 시작"""
        logger.info("브라우저 시작")
        
        self._playwright = await async_playwright().start()
        self._browser = await self._playwright.chromium.launch(
            headless=self.config.headless,
            args=['--disable-blink-features=AutomationControlled']
        )
        
        self._context = await self._browser.new_context(
            user_agent=self.config.user_agent,
            viewport={
                'width': self.config.viewport_width,
                'height': self.config.viewport_height
            },
            # 자동화 감지 방지
            extra_http_headers={
                'Accept-Language': 'ko-KR,ko;q=0.9,en-US;q=0.8,en;q=0.7',
            }
        )
        
        # 타임아웃 설정
        self._context.set_default_timeout(self.config.timeout)
        self._context.set_default_navigation_timeout(self.config.timeout)
        
        self._page = await self._context.new_page()
        
        logger.info("브라우저 시작 완료")
    
    async def close(self):
        """브라우저 종료"""
        logger.info("브라우저 종료")
        
        if self._page:
            await self._page.close()
        if self._context:
            await self._context.close()
        if self._browser:
            await self._browser.close()
        if self._playwright:
            await self._playwright.stop()
        
        logger.info("브라우저 종료 완료")
    
    async def get_page(self) -> Page:
        """현재 페이지 반환"""
        if not self._page:
            raise RuntimeError("브라우저가 시작되지 않았습니다. start()를 먼저 호출하세요.")
        return self._page
    
    async def new_page(self) -> Page:
        """새 페이지 생성"""
        if not self._context:
            raise RuntimeError("브라우저가 시작되지 않았습니다. start()를 먼저 호출하세요.")
        
        page = await self._context.new_page()
        logger.info("새 페이지 생성")
        return page
    
    @retry(
        stop=stop_after_attempt(3),
        wait=wait_exponential(multiplier=1, min=2, max=10),
        reraise=True
    )
    async def goto(self, url: str, wait_until: str = "networkidle") -> Page:
        """URL로 이동 (재시도 포함)"""
        page = await self.get_page()
        
        logger.info(f"페이지 이동: {url}")
        
        try:
            await page.goto(url, wait_until=wait_until, timeout=self.config.timeout)
            logger.info(f"페이지 로딩 완료: {url}")
            return page
        except Exception as e:
            logger.error(f"페이지 로딩 실패: {url}", error=str(e))
            raise
    
    async def wait_for_selector(
        self,
        selector: str,
        timeout: Optional[int] = None,
        state: str = "visible"
    ):
        """셀렉터 대기"""
        page = await self.get_page()
        timeout = timeout or self.config.timeout
        
        try:
            await page.wait_for_selector(selector, timeout=timeout, state=state)
        except Exception as e:
            logger.warning(f"셀렉터 대기 실패: {selector}", error=str(e))
            raise
    
    async def wait_for_load_state(self, state: str = "networkidle"):
        """페이지 로드 상태 대기"""
        page = await self.get_page()
        await page.wait_for_load_state(state, timeout=self.config.timeout)
    
    async def scroll_to_bottom(self, step: int = 500, delay: float = 0.5):
        """페이지 끝까지 스크롤"""
        page = await self.get_page()
        
        previous_height = await page.evaluate("document.body.scrollHeight")
        
        while True:
            # 스크롤 다운
            await page.evaluate(f"window.scrollBy(0, {step})")
            await asyncio.sleep(delay)
            
            # 새 높이 확인
            current_height = await page.evaluate("window.scrollHeight")
            current_scroll = await page.evaluate("window.pageYOffset + window.innerHeight")
            
            # 끝에 도달했는지 확인
            if current_scroll >= current_height:
                # 추가 컨텐츠 로딩 대기
                await asyncio.sleep(delay * 2)
                new_height = await page.evaluate("document.body.scrollHeight")
                
                # 더 이상 새 컨텐츠가 없으면 종료
                if new_height == previous_height:
                    break
                
                previous_height = new_height
    
    async def screenshot(self, path: str, full_page: bool = False):
        """스크린샷 저장"""
        page = await self.get_page()
        await page.screenshot(path=path, full_page=full_page)
        logger.info(f"스크린샷 저장: {path}")

File: src/core/browser.ts
import pRetry from 'p-retry';
import { Browser, BrowserContext, Page, chromium } from 'playwright';
import { randomDelay } from '../utils/delay.js';
import { getLogger } from '../utils/logger.js';
import { SiteConfig, defaultConfigValues } from './config.js';

const logger = getLogger('BrowserManager');

export class BrowserManager {
  private browser: Browser | null = null;
  private context: BrowserContext | null = null;
  private page: Page | null = null;
  private config: SiteConfig;

  // 병렬 처리용 추가 페이지들
  private extraPages: Page[] = [];

  constructor(config: Partial<SiteConfig> & { targetUrl: string }) {
    this.config = {
      ...defaultConfigValues,
      ...config,
    } as SiteConfig;
  }

  async start(): Promise<void> {
    logger.info('Starting browser...');

    this.browser = await chromium.launch({
      headless: this.config.headless,
      args: [
        '--disable-blink-features=AutomationControlled',
        '--no-sandbox',
        '--disable-setuid-sandbox',
        '--disable-dev-shm-usage',
        '--disable-accelerated-2d-canvas',
        '--no-first-run',
        '--no-zygote',
        '--disable-gpu',
      ],
    });

    this.context = await this.browser.newContext({
      viewport: {
        width: this.config.viewportWidth,
        height: this.config.viewportHeight,
      },
      userAgent: this.config.userAgent,
      locale: 'ko-KR',
      extraHTTPHeaders: {
        'Accept-Language': 'ko-KR,ko;q=0.9,en-US;q=0.8,en;q=0.7',
      },
    });

    await this.context.addInitScript(() => {
      Object.defineProperty(navigator, 'webdriver', {
        get: () => undefined,
      });
    });

    this.page = await this.context.newPage();

    logger.info({ headless: this.config.headless }, 'Browser started successfully');
  }

  async close(): Promise<void> {
    logger.info('Closing browser...');

    if (this.page) {
      await this.page.close().catch(() => {});
      this.page = null;
    }

    if (this.context) {
      await this.context.close().catch(() => {});
      this.context = null;
    }

    if (this.browser) {
      await this.browser.close().catch(() => {});
      this.browser = null;
    }

    logger.info('Browser closed');
  }

  getPage(): Page {
    if (!this.page) {
      throw new Error('Browser not started. Call start() first.');
    }
    return this.page;
  }

  getContext(): BrowserContext {
    if (!this.context) {
      throw new Error('Browser not started. Call start() first.');
    }
    return this.context;
  }

  async goto(url: string, options?: { waitUntil?: 'load' | 'domcontentloaded' | 'networkidle' }): Promise<void> {
    const page = this.getPage();
    // TODO: 다른 사이트도 통신사 선택 페이지 redirect 처리가 필요하다면 로직 변경 필요
    // 하이폰 특수 처리: 통신사 선택 세션이 필요함
    if (url.includes('hi-phone.kr') && !this.hiPhoneCarrierSelected) {
      await this.setupHiPhoneCarrier();
    }

    await pRetry(
      async () => {
        logger.debug({ url }, 'Navigating to URL');
        await page.goto(url, {
          waitUntil: options?.waitUntil || 'networkidle',
          timeout: this.config.timeout,
        });
        logger.debug({ url }, 'Navigation complete');
      },
      {
        retries: this.config.maxRetries,
        minTimeout: this.config.retryDelay * 1000,
        onFailedAttempt: (error) => {
          logger.warn(
            { url, attempt: error.attemptNumber, retriesLeft: error.retriesLeft },
            'Navigation failed, retrying...'
          );
        },
      }
    );
  }

  /**
   * 하이폰 통신사 선택 세션 설정
   * 하이폰은 통신사를 먼저 선택해야 리스팅/상세 페이지에 접근 가능
   * 쿠키 기반으로 세션 설정
   */
  private hiPhoneCarrierSelected = false;

  private async setupHiPhoneCarrier(): Promise<void> {
    const context = this.getContext();

    try {
      logger.info('Setting up hi-phone carrier cookies...');

      // 필수 쿠키 설정 (통신사 선택 세션용)
      // 핵심: my_agency 쿠키가 통신사 선택 상태를 결정함
      const now = Date.now();
      await context.addCookies([
        {
          name: 'my_agency',
          value: 'SKT',  // 기본 통신사로 SKT 선택
          domain: 'hi-phone.kr',
          path: '/',
          expires: Math.floor(Date.now() / 1000) + 86400 * 365, // 1년
        },
        {
          name: 'cartId',
          value: `scraper_cart_${now}`,
          domain: 'hi-phone.kr',
          path: '/',
        },
        {
          name: 'smipUserToken',
          value: `${now.toString(16)}-scraper-token`,
          domain: 'hi-phone.kr',
          path: '/',
          expires: Math.floor(Date.now() / 1000) + 86400 * 365, // 1년
        },
        {
          name: 'smtg_cKey',
          value: now.toString(),
          domain: '.hi-phone.kr',
          path: '/',
          expires: Math.floor(Date.now() / 1000) + 86400,
        },
        {
          name: 'smtg_fsID',
          value: '1',
          domain: '.hi-phone.kr',
          path: '/',
        },
        {
          name: 'smtg_sAd',
          value: '0',
          domain: '.hi-phone.kr',
          path: '/',
        },
        {
          name: 'smtg_sKey',
          value: now.toString(),
          domain: '.hi-phone.kr',
          path: '/',
        },
        {
          name: 'smtg_vTime',
          value: Math.floor(now / 1000).toString(),
          domain: '.hi-phone.kr',
          path: '/',
          expires: Math.floor(Date.now() / 1000) + 86400 * 90,
        },
      ]);

      this.hiPhoneCarrierSelected = true;
      logger.info('Hi-phone cookies set successfully');

    } catch (error) {
      logger.warn({ error }, 'Hi-phone cookie setup failed, trying page-based selection');

      // 쿠키 설정 실패 시 페이지 기반 선택 시도
      await this.setupHiPhoneCarrierViaPage();
    }
  }

  /**
   * 페이지 기반 통신사 선택 (쿠키 방식 실패 시 fallback)
   */
  private async setupHiPhoneCarrierViaPage(): Promise<void> {
    const page = this.getPage();

    try {
      await page.goto('https://hi-phone.kr/index.php', {
        waitUntil: 'domcontentloaded',
        timeout: 30000,
      });

      await randomDelay(1, 2);

      // SKT 클릭 시도
      try {
        await page.click('text=SKT', { timeout: 5000 });
        await randomDelay(1, 2);
        logger.info('Clicked SKT via page');
      } catch {
        logger.warn('Could not click SKT on hi-phone');
      }

      this.hiPhoneCarrierSelected = true;
    } catch (error) {
      logger.warn({ error }, 'Hi-phone page-based setup failed');
      this.hiPhoneCarrierSelected = true;
    }
  }

  async screenshot(options?: { fullPage?: boolean }): Promise<Buffer> {
    const page = this.getPage();
    return await page.screenshot({
      fullPage: options?.fullPage ?? false,
      type: 'png',
    });
  }

  async screenshotBase64(options?: { fullPage?: boolean }): Promise<string> {
    const buffer = await this.screenshot(options);
    return buffer.toString('base64');
  }

  async getHtml(): Promise<string> {
    const page = this.getPage();
    return await page.content();
  }

  async getSimplifiedHtml(): Promise<string> {
    const page = this.getPage();

    return await page.evaluate(() => {
      const clone = document.body.cloneNode(true) as HTMLElement;

      const removeElements = ['script', 'style', 'noscript', 'iframe', 'svg', 'path'];
      removeElements.forEach((tag) => {
        clone.querySelectorAll(tag).forEach((el) => el.remove());
      });

      const removeAttrs = [
        'style',
        'onclick',
        'onload',
        'onerror',
        'onmouseover',
        'onmouseout',
        'onfocus',
        'onblur',
      ];
      clone.querySelectorAll('*').forEach((el) => {
        removeAttrs.forEach((attr) => el.removeAttribute(attr));
      });

      return clone.innerHTML
        .replace(/\s+/g, ' ')
        .replace(/>\s+</g, '><')
        .trim();
    });
  }

  async waitForSelector(selector: string, timeout?: number): Promise<void> {
    const page = this.getPage();
    await page.waitForSelector(selector, { timeout: timeout || this.config.timeout });
  }

  async click(selector: string): Promise<void> {
    const page = this.getPage();
    await page.click(selector);
  }

  async applyRandomDelay(): Promise<void> {
    await randomDelay(this.config.delayMin, this.config.delayMax);
  }

  async evaluate<T>(fn: () => T | Promise<T>): Promise<T> {
    const page = this.getPage();
    return await page.evaluate(fn);
  }

  /**
   * 병렬 처리용 새 페이지 생성
   */
  async createExtraPage(): Promise<Page> {
    const context = this.getContext();
    const newPage = await context.newPage();
    this.extraPages.push(newPage);
    logger.debug({ totalPages: this.extraPages.length + 1 }, 'Created extra page for parallel processing');
    return newPage;
  }

  /**
   * 여러 개의 추가 페이지 생성
   */
  async createExtraPages(count: number): Promise<Page[]> {
    const pages: Page[] = [];
    for (let i = 0; i < count; i++) {
      const page = await this.createExtraPage();
      pages.push(page);
    }
    return pages;
  }

  /**
   * 추가 페이지들 정리
   */
  async closeExtraPages(): Promise<void> {
    for (const page of this.extraPages) {
      try {
        await page.close();
      } catch {
        // Ignore close errors
      }
    }
    this.extraPages = [];
    logger.debug('Closed all extra pages');
  }

  /**
   * 특정 페이지로 URL 이동 (병렬 처리용)
   */
  async gotoWithPage(targetPage: Page, url: string, options?: { waitUntil?: 'load' | 'domcontentloaded' | 'networkidle' }): Promise<void> {
    // 하이폰 특수 처리
    if (url.includes('hi-phone.kr') && !this.hiPhoneCarrierSelected) {
      await this.setupHiPhoneCarrier();
    }

    await pRetry(
      async () => {
        logger.debug({ url }, 'Navigating to URL (parallel)');
        await targetPage.goto(url, {
          waitUntil: options?.waitUntil || 'networkidle',
          timeout: this.config.timeout,
        });
        logger.debug({ url }, 'Navigation complete (parallel)');
      },
      {
        retries: this.config.maxRetries,
        minTimeout: this.config.retryDelay * 1000,
        onFailedAttempt: (error) => {
          logger.warn(
            { url, attempt: error.attemptNumber, retriesLeft: error.retriesLeft },
            'Navigation failed, retrying...'
          );
        },
      }
    );
  }

}
